numpy/core/src/umath/reduction.c

```diff
@@ -313,12 +325,40 @@ PyUFunc_ReduceWrapper(PyArrayMethod_Context *context,
      * Initialize the result to the reduction unit if possible,
      * otherwise copy the initial values and get a view to the rest.
      */
-    if (identity != Py_None) {
-        if (PyArray_FillWithScalar(result, identity) < 0) {
+    if (initial != NULL && initial != Py_None) {
+        /*
+         * User provided an `initial` value and it is not `None`.
+         * NOTE: It may make sense to accept array-valued `initial`,
+         *       this would subtly (but rarely) change the coercion of
+         *       `initial`.  But it would be perfectly fine otherwise.
+         */
+        if (PyArray_FillWithScalar(result, initial) < 0) {
+            goto fail;
+        }
+    }
+    else if (identity_buf != NULL && (  /* cannot fill for `initial=None` */
+                identity_flags & NPY_METH_ITEM_IS_IDENTITY ||
+                (empty_reduce && identity_flags & NPY_METH_ITEM_IS_DEFAULT))) {
+        /* Loop provided an identity or default value, assign to result. */
+        int ret = raw_array_assign_scalar(
```

Comment:
It seems a bit confusing that this uses `raw_array_assign_scalar` rather than `PyArray_FillWithScalar`. Why are the built-in identities and user-provided ones treated differently? Is it still possible for the `get_identity` function to return a `PyObject` for filling with `dtype=object`? Part of the reason for wondering about always returning something reference counted elsewhere...

Reply:
This has become irrelevant; I also see now that the tests done in `PyArray_FillWithScalar` are not needed any more. So, this can probably be resolved.